### PR TITLE
Enter highscore name in a floating window

### PR DESCRIPTION
**Pull request by dospuntos**
This pull request addresses Issue #21 and introduces a floating window for entering highscore name. It also displays the point difference when there is a previous highscore.

### Changes Made
- Added a new window for entering highscore name
- Implemented the display of the point difference for previous highscores (if available)
- Updated Game.cpp to save previous record and username
- Updated the screenshot

Please review and let me know if any changes are needed.

### CODE REVIEW

File: HighscoreWindow.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		BTextControl("NameInput", NULL, NULL, new BMessage(H2048_NAME_ENTERED));
+	fInputBox->SetText(B_TRANSLATE("Anonymous"));
+	fInputBox->SetAlignment(B_ALIGN_HORIZONTAL_CENTER,B_ALIGN_HORIZONTAL_CENTER);
+	fInputBox->MakeFocus();
+	
+	BString previousWinner;
+	previousWinner << B_TRANSLATE("This is the first record!");
+	if (oldHighscorer[0]) {
+		// Replace with previous winner and score difference
+		previousWinner = B_TRANSLATE("You've beaten the previous record holder\n"
```
Comment (humdingerb) [inline HighscoreWindow.cpp:57]: I think the %oldHighscorer should be between commas:
```
"You've beaten the previous record holder,\n"
"%oldHighscorer%, by %deltaScore%."
```

Or maybe even simpler:
```
"You've beaten "%oldHighscorer%'s\n"
"old record by %deltaScore%."
```

Your call.

**Review by humdingerb — COMMENTED** [review]
Looks good, until you enter a very long user name. Try with "TheGuyWithAVeryLongUsernameThatPlaysToWin" :)

Maybe just limit the input to, say, 25 characters?

And maybe squash your commits...

File: HighscoreWindow.cpp (hunk 1 above)
Comment (dospuntos) [reply HighscoreWindow.cpp]: OK, I updated with the first option (and squashed the commits).

I'm unsure how to limit the input length, any hints?

File: Game.cpp [hunk 2]
```diff
@@ -112,6 +112,8 @@ Game::newGame()
 {
 	fScore = 0;
 	fCanUndo = false;
+	fPreviousHighscore = fScore_Highest;
+	memcpy(fPreviousUsername, fUsername, sizeof(char) * 32);
```
Comment (korli) [inline Game.cpp:116]:
```suggestion
	strlcpy(fPreviousUsername, fUsername, sizeof(fPreviousUsername));
```

File: HighscoreWindow.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+class BStringView;
+class BTextControl;
+
+enum
+{
+	H2048_NAME_ENTERED = '48NE',
+};
+
+class HighscoreWindow : public BWindow{
```
Comment (korli) [inline HighscoreWindow.h:22]: following style in Game.h
```suggestion
class HighscoreWindow : public BWindow
{
```

File: HighscoreWindow.cpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+			.AddStrut(2)
+			.Add(inputlabel)
+			.Add(fInputBox)
+			.Add(saveName)
+			.AddGlue()
+			.End();
+			
+	this->ResizeToPreferred();
+}
+
```
Comment (korli) [inline HighscoreWindow.cpp:88]:
```suggestion


```

File: HighscoreWindow.cpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	{
+		case H2048_NAME_ENTERED:
+		{
+			// Set a default value if empty
+			if (!fInputBox->Text()[0])
+				fInputBox->SetText(B_TRANSLATE("Anonymous"));
+
+			BMessage *req = new BMessage(H2048_NAME_REQUESTED);
+			req->AddString("playername", fInputBox->Text());
+			fMessenger.SendMessage(req);
```
Comment (korli) [inline HighscoreWindow.cpp:114]:
```suggestion
			BMessage req(H2048_NAME_REQUESTED);
			req.AddString("playername", fInputBox->Text());
			fMessenger.SendMessage(&req);
```

File: WindowBoard.cpp [hunk 6]
```diff
@@ -258,7 +247,7 @@ GameWindow::FrameResized(float width,
 {
 	// We don't want the user to scale the window so small that
 	// there's no space for the buttons.
-	if (width < defaultWidth) {
+	if (width < defaultWidth && height < defaultHeight) {
```
Comment (korli) [inline WindowBoard.cpp:250]: Shouldn't be || ?

File: WindowBoard.h [hunk 7]
```diff
@@ -1,15 +1,18 @@
...
  * Copyright (c) 2015 Markus Himmel
+ * Copyright 2022, Harshit Sharma <harshits908@gmail.com>
  * This file is distributed under the terms of the MIT license
  */
 
 #ifndef WINDOW_BOARD_H
 #define WINDOW_BOARD_H
 
 #include "GameBoard.h"
+#include "HighscoreWindow.h"
```
Comment (korli) [inline WindowBoard.h:11]: Why is it needed?

File: HighscoreWindow.cpp (hunk 1 above)
Comment (humdingerb) [reply HighscoreWindow.cpp]: > I'm unsure how to limit the input length, any hints?

Either, you silently truncate the string when the user is done entering.
Or, you can use the BTextControl's SetModificationMessage() that sends a message for every letter entered. Then check with every letter if the limit was reached and remove the last entered letter. That should prevent the user to enter any more letters. I think.
The second option should be enough to signal the input limit to the user. I wouldn't pop up an alert or something. It's obvious enough.

File: WindowBoard.h (hunk 7 above)
Comment (dospuntos) [reply WindowBoard.h]: Thanks for your reviews, @korli, I've added your suggestions to the pull request.
I moved this #include to `WindowBoard.cpp`.

File: HighscoreWindow.cpp (hunk 1 above)
Comment (dospuntos) [reply HighscoreWindow.cpp]: Thanks, I added a SetModificationMessage() to make sure the input is limited to 25 characters, and applied some other recommendations from @korli . Please review when you have a chance.

File: HighscoreWindow.cpp [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	BStringView* inputlabel = new BStringView("inputlabel",
+										B_TRANSLATE("Please enter your name:"));
+	inputlabel->SetAlignment(B_ALIGN_HORIZONTAL_CENTER);
+	inputlabel->SetFont(be_bold_font);
+
+	fInputBox = new
+		BTextControl("NameInput", NULL, NULL, new BMessage(H2048_NAME_ENTERED));
+	fInputBox->SetText(B_TRANSLATE("Anonymous"));
+	fInputBox->SetAlignment(B_ALIGN_HORIZONTAL_CENTER,B_ALIGN_HORIZONTAL_CENTER);
+	fInputBox->MakeFocus();
```
Comment (humdingerb) [inline HighscoreWindow.cpp:51]: If you move this line to the very end of this constructor, the text in the text control actually gets selected and the user can start typing.

File: HighscoreWindow.cpp [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+			BMessage req(H2048_NAME_REQUESTED);
+			req.AddString("playername", fInputBox->Text());
+			fMessenger.SendMessage(&req);
+			Quit();
+			break;
+		}	
+		case H2048_CHANGE_NAME:
+		{
+			BString name = fInputBox->Text();
+			if (name.CountChars() > 25)
```
Comment (humdingerb) [inline HighscoreWindow.cpp:123]: While this works, the text control losing focus means that the user cannot just backspace a bit and use a shorter user name.

I suggest this, which leaves the input carret at the end of the text field:

```
		case H2048_CHANGE_NAME:
		{
			BString name = fInputBox->Text();
			int32 count = name.CountChars();
			if (count > 25)
				fInputBox->TextView()->Delete(count - 1, count);
			break;
		}
```

**Review by humdingerb — COMMENTED** [review]
Also, have a look for stray whitespaces, there are a few of it...
I use "Koder" with activated "Highlight trailing whitespace" which make them stick out.

**Comment by dospuntos** [comment]
@humdingerb - I've removed stray white spaces. The input is now focused when the window opens, and won't lose focus if the 25-character limit is reached. Thanks for the advice 👍

**Comment by humdingerb** [comment]
Very good! When @korli gives his thumbs up, we merge.

**Comment by korli** [comment]
Looks good enough for me!

**Pull request merged**